llvm/lib/SandboxIR/SandboxIR.cpp

```diff
@@ -1150,6 +1150,22 @@ Value *PHINode::hasConstantValue() const {
   llvm::Value *LLVMV = cast<llvm::PHINode>(Val)->hasConstantValue();
   return LLVMV != nullptr ? Ctx.getValue(LLVMV) : nullptr;
 }
+void PHINode::replaceIncomingBlockWith (const BasicBlock *Old, BasicBlock *New) {
+  assert(New && Old && "Sandbox IR PHI node got a null basic block!");
+  for (unsigned Op = 0,
+            NumOps = cast<llvm::PHINode>(Val)->getNumOperands(); Op != NumOps; ++Op)
+    if (getIncomingBlock(Op) == Old)
+      setIncomingBlock(Op, New);
+}
+void PHINode::removeIncomingValueIf(function_ref< bool(unsigned)> Predicate,
+                                    bool DeletePHIIfEmpty) {
+  // Avoid duplicate tracking by going through this->removeIncomingValue here at
+  // the expense of some performance. Copy PHI::removeIncomingValueIf more
+  // directly if performance becomes an issue.
+  for (unsigned Idx = 0; Idx < getNumIncomingValues(); ++Idx)
```

Comment:
The `getNumIncomingValues()` could be seen as a coding standards violation:

https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

Reply:
But I think in this case it is needed, as it is removing incoming values while iterating.

Reply:
Fine.

Reply:
Hmm is this working correctly though when we delete an incoming value? Shouldn't `Idx` not get incremented if `Predicate(Idx)` returns true?

Reply:
When you said that the `getNumIncomingValues()` is needed, I thought of a while loop.
```
unsigned Idx = 0;
while (getIncomingValues() > 0) {
 if (Predicate(Idx))
      removeIncomingValue(Idx);
  else
  ++Idx;
}
```

Reply:
I don't think that quite works because we usually aren't removing all incoming values, so getIncomingValues will still be greater than zero even after iterating over everything. 

```
unsigned Idx = 0;
while (Idx < getIncomingValues()) {
   if (Predicate(Idx))
       removeIncomingValue(Idx);
   else
       ++Idx;
}
```

Or, if we really do care about avoiding the function call every time through the loop:

```
unsigned Idx = 0;
unsigned LastIdx = getIncomingValues();
while (Idx < LastIdx) {
   if (Predicate(Idx))
       removeIncomingValue(Idx);
       --LastIdx;
   else
       ++Idx;
}
```